ADAL/src/ADAuthenticationResult+Internal.m

```diff
@@ -155,20 +161,77 @@ + (ADAuthenticationResult*)resultForBrokerErrorResponse:(NSDictionary*)response
     if (strErrorCode && ![strErrorCode isEqualToString:@"0"])
     {
         errorCode = [strErrorCode integerValue];
     }
+
     
+
+    if (errorCode == AD_ERROR_SERVER_PROTECTION_POLICY_REQUIRED)
+    {
+        // In the case where Intune App Protection Policies are required, the broker may send back the Intune MAM Resource token
+        userInfo = [[NSMutableDictionary alloc] initWithCapacity:3];
+        if (response[@"suberror"])
+        {
+            [userInfo setValue:response[@"suberror"] forKey:@"suberror"];
+        }
+
+        ADAuthenticationError* intuneTokenError = nil;
+        if (response[BROKER_INTUNE_HASH_KEY] && response[BROKER_INTUNE_RESPONSE_KEY])
+        {
+            NSDictionary* intuneTokenResponse = [ADHelpers decryptBrokerResponse:@{BROKER_RESPONSE_KEY:response[BROKER_INTUNE_RESPONSE_KEY],
+                                                                                   BROKER_HASH_KEY:response[BROKER_INTUNE_HASH_KEY],
+                                                                                   BROKER_MESSAGE_VERSION:response[BROKER_MESSAGE_VERSION] ? response[BROKER_MESSAGE_VERSION] : @1}
+                                                                   correlationId:correlationId
+                                                                           error:&intuneTokenError];
+
+            ADAuthenticationResult* intuneTokenResult = [[ADTokenCacheItem new] processTokenResponse:intuneTokenResponse
+                                                                                         fromRefresh:NO
+                                                                                requestCorrelationId:intuneTokenResponse[OAUTH2_CORRELATION_ID_RESPONSE]];
+
+            if (intuneTokenResult)
+            {
+                [userInfo setValue:intuneTokenResult.tokenCacheItem.userInformation.userId forKey:@"userID"];
+            }
+
+            NSDictionary* resumeDictionary = [[NSUserDefaults standardUserDefaults] objectForKey:kAdalResumeDictionaryKey];
+            if (!resumeDictionary)
+            {
+                AD_LOG_WARN(correlationId, @"Failed to cache Intune token, no resume state found in NSUserDefaults");
+            }
+            else
+            {
+                NSString* keychainGroup = resumeDictionary[@"keychain_group"];
+                id<ADTokenCacheDataSource> cache = [ADKeychainTokenCache keychainCacheForGroup:keychainGroup];
+                ADTokenCacheAccessor* cacheAccessor = [[ADTokenCacheAccessor alloc] initWithDataSource:cache
+                                                                                             authority:intuneTokenResult.tokenCacheItem.authority];
+
+                [cacheAccessor updateCacheToResult:intuneTokenResult cacheItem:nil refreshToken:nil context:nil];
+            }
+        }
+
+        if (intuneTokenError)
+        {
+            [userInfo setValue:intuneTokenError forKey:@"intuneTokenError"];
+        }
+    }
+
```

Comment:
I am thinking that if we should keep the result caching and response decryption to `ADAuthenticationRequest+Broker.m`, such that ADAuthenticationResult doesn't need to know anything about caching. The other side effect of this solution is that the decryption part of code is also duplicated in two places. I actually like your previous solution better :P
What do you think @oldalton ?